backend/arm64/emit.mlp

```diff
@@ -987,21 +981,139 @@ let emit_load_literal dst lbl =
     `	ldr	{emit_reg dst}, [{emit_reg reg_tmp1}, #:lo12:{emit_label lbl}]\n`
   end
 
-
-let move src dst =
-  if src.loc <> dst.loc then begin
-    match (src, dst) with
-    | {loc = Reg _; typ = Float}, {loc = Reg _} ->
-      `	fmov	{emit_reg dst}, {emit_reg src}\n`
-    | {loc = Reg _}, {loc = Reg _} ->
-      `	mov	{emit_reg dst}, {emit_reg src}\n`
-    | {loc = Reg _}, {loc = Stack _} ->
-      `	str	{emit_reg src}, {emit_stack dst}\n`
-    | {loc = Stack _}, {loc = Reg _} ->
-      `	ldr	{emit_reg dst}, {emit_stack src}\n`
-    | _ ->
-      assert false
-  end[@@warning "-4"]
+let move (src : Reg.t) (dst : Reg.t) =
+  let distinct = not (Reg.same_loc src dst) in
+  if distinct then
+  match src.typ, src.loc, dst.typ, dst.loc with
+  | Float, Reg _, Float, Reg _
+  | Float32, Reg _, Float32, Reg _
+  | (Vec128|Valx2), Reg _, (Vec128|Valx2), Reg _ ->
+      DSL.ins I.FMOV [| DSL.emit_reg dst; DSL.emit_reg src |]
+  | (Int | Val | Addr), Reg _, (Int | Val | Addr), Reg _ ->
+      DSL.ins I.MOV [| DSL.emit_reg dst; DSL.emit_reg src |]
+  | _, Reg _, _, Stack _ ->
+     `	str	{emit_reg src}, {emit_stack dst}\n`
+  | _, Stack _, _, Reg _ ->
+     `	ldr	{emit_reg dst}, {emit_stack src}\n`
+  | _, Stack _, _, Stack _ ->
+      Misc.fatal_errorf
+      "Illegal move between registers (%a to %a)\n"
+      Printreg.reg src Printreg.reg dst
+  | _, Unknown, _, (Reg _ | Stack _ | Unknown)
+  | _, (Reg _ | Stack _), _, Unknown ->
+      Misc.fatal_errorf
+      "Illegal move with an unknown register location (%a to %a)\n"
+      Printreg.reg src Printreg.reg dst
+  | (Float | Float32 | Vec128 | Int | Val | Addr | Valx2), (Reg _), _, _ ->
+      Misc.fatal_errorf
+       "Illegal move between registers of differing types (%a to %a)\n"
+       Printreg.reg src Printreg.reg dst
+
+let emit_reinterpret_cast (cast : Cmm.reinterpret_cast) i =
+    let src = i.arg.(0) in
+    let dst = i.res.(0) in
+    let distinct = not (Reg.same_loc src dst) in
+    match cast with
+    | Int64_of_float ->
+      DSL.check_reg Float src;
+      DSL.ins I.FMOV [| DSL.emit_reg dst; DSL.emit_reg src |]
+    | Float_of_int64 ->
+      DSL.check_reg Float dst;
+      DSL.ins I.FMOV [| DSL.emit_reg dst; DSL.emit_reg src |]
+    | Float32_of_int32 ->
+      DSL.check_reg Float32 dst;
+      DSL.ins I.FMOV [| DSL.emit_reg dst; DSL.emit_reg_w src |]
+    | Int32_of_float32 ->
+      DSL.check_reg Float32 src;
+      DSL.ins I.FMOV [| DSL.emit_reg_w dst; DSL.emit_reg src |]
+    | Float32_of_float ->
+       if distinct then (
+        DSL.check_reg Float src;
+        DSL.check_reg Float32 dst;
+        (* CR melse: I think this might be broken - maybe we need to do fmov dN,
+           dM without referring to single-precision registers? *)
+          DSL.ins I.FMOV [| DSL.emit_reg dst; DSL.emit_reg src |])
```

Comment:
yeah apparently fmov 32<->64 is a static cast

Reply:
fixed to use FCVT not FMOV and removed the addressed CR.

Reply:
Wrong way! it needs fmov 64->64

Reply:
We should probably also add some tests for reinterpret float/float32

Reply:
oops, i got it mixed up with static cast, thanks for the catch. the tests will be enabled in a follow-up pr.